primitives/utility/src/lib.rs

```diff
@@ -51,3 +58,199 @@ impl<T: UpwardMessageSender, W: WrapVersion> SendXcm for ParentAsUmp<T, W> {
+/// Alternatively we could just return payment in the aforementioned case
+pub struct TakeFirstAssetTrader<
+	AccountId,
+	FeeCharger: ChargeWeightInFungibles<AccountId, ConcreteAssets>,
+	Matcher: MatchesFungibles<ConcreteAssets::AssetId, ConcreteAssets::Balance>,
+	ConcreteAssets: fungibles::Mutate<AccountId> + fungibles::Transfer<AccountId> + fungibles::Balanced<AccountId>,
+	HandleRefund: TakeRevenue,
+>(
+	Option<AssetTraderRefunder>,
+	PhantomData<(AccountId, FeeCharger, Matcher, ConcreteAssets, HandleRefund)>,
+);
+impl<
+		AccountId,
+		FeeCharger: ChargeWeightInFungibles<AccountId, ConcreteAssets>,
+		Matcher: MatchesFungibles<ConcreteAssets::AssetId, ConcreteAssets::Balance>,
+		ConcreteAssets: fungibles::Mutate<AccountId>
+			+ fungibles::Transfer<AccountId>
+			+ fungibles::Balanced<AccountId>,
+		HandleRefund: TakeRevenue,
+	> WeightTrader
+	for TakeFirstAssetTrader<AccountId, FeeCharger, Matcher, ConcreteAssets, HandleRefund>
+{
+	fn new() -> Self {
+		Self(None, PhantomData)
+	}
+	// We take first multiasset
+	// Check whether we can convert fee to asset_fee (is_sufficient, min_deposit)
+	// If everything goes well, we charge.
+	fn buy_weight(
+		&mut self,
+		weight: Weight,
+		payment: xcm_executor::Assets,
+	) -> Result<xcm_executor::Assets, XcmError> {
+		log::trace!(target: "xcm::weight", "TakeFirstAssetTrader::buy_weight weight: {:?}, payment: {:?}", weight, payment);
+
+		// Make sure we dont enter twice
+		if self.0.is_some() {
+			return Err(XcmError::NotWithdrawable)
+		}
+
+		// We take the very first multiasset from payment
+		let multiassets: MultiAssets = payment.clone().into();
+
+		// Take the first multiasset from the selected MultiAssets
+		let first = multiassets.get(0).ok_or(XcmError::AssetNotFound)?;
+
+		// Get the local asset id in which we can pay for fees
+		let (local_asset_id, _) =
+			Matcher::matches_fungibles(&first).map_err(|_| XcmError::AssetNotFound)?;
+
+		// Calculate how much we should charge in the asset_id for such amount of weight
+		let asset_balance = FeeCharger::charge_weight_in_fungibles(local_asset_id, weight)?;
+
+		match first {
+			// Not relevant match, as matches_fungibles should have already verified this above
+			MultiAsset {
+				id: xcm::latest::AssetId::Concrete(location),
+				fun: Fungibility::Fungible(_),
+			} => {
+				// Convert asset_balance to u128
+				let u128_amount: u128 = asset_balance.try_into().map_err(|_| XcmError::Overflow)?;
+				// Construct required payment
+				let required: MultiAsset = (Concrete(location.clone()), u128_amount).into();
+				// Substract payment
+				let unused =
+					payment.checked_sub(required.clone()).map_err(|_| XcmError::TooExpensive)?;
+
+				// record weight, balance, multilocation and assetId
+				self.0 = Some(AssetTraderRefunder {
+					weight_outstanding: weight,
+					outstanding_concrete_asset: required,
+				});
+				Ok(unused)
+			},
+			_ => return Err(XcmError::TooExpensive),
```

Comment:
I think we should also use `AssetNotFound` here as well.

Ideally, there should be some form of static assertion in this struct here that forbids anything other than a concrete asset type from being assigned as the 4th type parameter, there might be a way to do this but I'm currently blanking out right now...

Reply:
I finally reworked the logic so that this match isnt needed